docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,361 @@
+
+```yaml
+build:
+  artifacts:
+  - image: builder-image
+  - image: run-image
+  - image: skaffold-buildpacks
+    buildpacks:
+      builder: "{{ .BUILDER_IMAGE }}"
+      runImage: "{{ .RUN_IMAGE }}"
+    requires:
+      - image: builder-image
+        alias: BUILDER_IMAGE
+      - image: run-image
+        alias: RUN_IMAGE
+```
+
+If there are any additional images in the `required` section it only enforces that they get built prior to the current image. However, the buildpacks builder cannot really reference them in any other way.
+
+### Jib builder
+
+The Jib builder supports [changing the base image](https://cloud.google.com/java/getting-started/jib#base-image). We allow substituting the `required` image aliases with the image references in the specific builder `args` section.
+This will allow overriding the `jib.from.image` property that sets the base image.
+
+For Maven:
+
+```yaml
+build:
+  artifacts:
+  - image: base-image
+  - image: test-jib-maven
+    jib:
+      type: maven
+      args: 
+      - -Djib.from.image=docker://{{ .BASE_IMAGE }}
+    requires:
+      - image: base-image
+        alias: BASE_IMAGE
+```
+
+where `jib.from.image` is added appropriately the `pom.xml`
+
+Similarly, for Gradle:
+
+```yaml
+build:
+  artifacts:
+  - image: base-image
+  - image: test-jib-gradle
+    jib:
+      type: gradle
+      args: 
+      - -Djib.from.image=registry://{{ .BASE_IMAGE }}
+    requires:
+      - image: base-image
+        alias: BASE_IMAGE
+```
+
+where `jib.from.image` is added to the Gradle configuration.
+
+### Bazel builder
+
+The bazel builder doesn't support referencing images directly. Also, it natively supports setting up nested builds. We allow defining `required` artifacts even though they can't be referenced by the builder. We do this to give the user a way of ordering these builds; with the future work around [Skaffold Hooks](https://github.com/GoogleContainerTools/skaffold/issues/1441) there might be some usecases where pre and post build scripts might want a certain ordering of builds. 
+
+## Builder interfaces
+There are two builder abstractions -- in [build.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/build.go#L37) and [parallel.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go#L34)
+
+```go
+type Builder interface {
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+}
+```
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
+```
+The first describes a builder for a list of artifacts. There are a few implementations like `cache`, `local` and `cluster` builders.
```

Comment:
Maybe we should rename this from `Builder` to `Platform` or something.

Reply:
discussed offline. Resolving.